---
title: "Linear model for transcription expression data on multiple cancers - using TCGA non-core splicing factors"
output: html_notebook
---

Descritpion: linear model to identify differential expression between cell lines maintaining an unaltered spliceosome and cell lines with mutated splicing factors. Splicing factor list from TCGA. 

Load libraries
```{r}
library(tidyverse)
library(edgeR)
library(ggpubr)
library(dplyr)
library(readr)
library(biomaRt)
library(data.table)
select = dplyr::select
rename = dplyr::rename
filter = dplyr::filter
```

source R profile. Memory was set to 500000.
```{r}
Sys.setenv("R_ENVIRON_USER"='/Users/castilln/.Renviron')
Sys.getenv("R_ENVIRON_USER")

```

Set working directory
```{r}
setwd("/Users/castilln/Desktop/thesis/localdata")
```


Load data
```{r}
transcripts <- read_csv("/Users/castilln/Desktop/thesis/localdata/depmap/CCLE_RNAseq_transcripts.csv")

#METADATA
sample_info <- read_csv("/Users/castilln/Desktop/thesis/localdata/depmap/sample_info.csv") #metadata

#MUTATIONS AND SAMPLE INFO
CCLE_mutations <- fread("/Users/castilln/Desktop/thesis/localdata/depmap/CCLE_info", header = TRUE) 

#LIST OF MUTATIONS FROM CCLE MUTATIONS THAT ONLY AFFECT THE CORE SPLICEOSOME
mutations_spliceosome = fread("spliceosome/tcga_non_core_spliceosome_ccle_mutations.csv")
```

Contingency table - Annotation of cell lines with spliceosome mutations
```{r}
cell_lines_list = 
  CCLE_mutations %>% 
  select(stripped_cell_line_name, primary_disease, DepMap_ID) 

cell_lines_list_mutated = 
 cell_lines_list %>% 
   mutate(spliceosome_mutated = 
           case_when(
            cell_lines_list$stripped_cell_line_name %in% mutations_spliceosome$stripped_cell_line_name ~ "YES", #if the mutation is in a gene from the spliceosome: 1 
            !cell_lines_list$stripped_cell_line_name %in% mutations_spliceosome$stripped_cell_line_name ~ "NO")) %>% 
  distinct()

cell_lines_list_mutated %>% 
  distinct() %>%
  group_by(primary_disease,spliceosome_mutated) %>%
  tally() 

table(distinct(cell_lines_list_mutated)$primary_disease,distinct(cell_lines_list_mutated)$spliceosome_mutated)

spliceosome_mutated_ID = 
  cell_lines_list_mutated %>% 
  select(DepMap_ID, spliceosome_mutated) %>% 
  distinct() 
```

```{r}
#RENAME FIRST VARIABLE 
transcripts = 
  transcripts %>% 
  rename("DepMap_ID" = "X1")

#JOIN TRANSCRIPT INFO WITH METADATA
long_transcripts_info = 
  transcripts %>% 
  pivot_longer(cols=-DepMap_ID, names_to = "Gene", values_to = "log_TPM") %>% 
  left_join(cell_lines_list_mutated, by = "DepMap_ID")

head(long_transcripts_info)
```

## Pre-processing: quality check. 
Let's keep only those cancers with a relatively significant number of cell lines w/o mutations in the spliceosome:
```{r}
no_keep <- c("Adrenal Cancer", "Cervical Cancer", "Embryonal Cancer", "Non-Cancerous", "Teratoma")

#define notin operator
`%notin%` <- Negate(`%in%`)

df_deseq = 
long_transcripts_info %>% 
  filter(primary_disease %notin% no_keep)

df_deseq %>% 
  filter(spliceosome_mutated == "NO") %>% 
  select(DepMap_ID) %>% 
  distinct() 
```

Filter out those genes whose median for expression is lower than 1 across all samples
```{r}
median_counts =
  df_deseq %>% 
  group_by(Gene) %>% 
  dplyr::mutate(median = median(log_TPM, na.rm=TRUE)) %>% 
  filter(median > 1)
```

Create a new column for ensembl transcript ID 
```{r}
median_counts = 
  median_counts %>%
    separate(col = "Gene", into = c("Hugo_Symbol", "ensembl_id"), sep = "\\s") 
   
median_counts$ensembl_id <- gsub("\\(|\\)","",as.character(median_counts$ensembl_id))

counts_lm = 
  median_counts %>% 
  select(Hugo_Symbol, ensembl_id, stripped_cell_line_name, primary_disease, spliceosome_mutated, log_TPM)

##CHECK THAT THERE ARE NO DUPLICATES
#counts_lm %>% 
#distinct() %>% 
#dim()
```

```{r}
head(counts_lm)
```

# QQ plot
```{r}
theme_set(
  theme_minimal() +
    theme(legend.position = "top")
  )

##PLOT 
ggplot(median_counts, aes(sample = log_TPM)) + 
  facet_wrap(facets = "primary_disease") +
  stat_qq() + 
  stat_qq_line() 
```

# Linear model

Fit directly to linear model
```{r}
library(broom)
lm_res =
  counts_lm %>% 
  ##NEST BY TRANSCRIPT
  nest(-ensembl_id) %>% 
  #LM - DESIGN
  mutate(model = map(data, ~lm(log_TPM ~ spliceosome_mutated + primary_disease, data = .))) %>% 
  #TIDY THE MODEL
  mutate(tidied = map(model, tidy)) %>%
  unnest(tidied) %>% 
  select(-c("data", "model"))

```

Get list of ensemnl transcript IDs with correspondent Hugo_Symbols 
```{r}
symbols_list = 
  median_counts %>% 
  select(c("Hugo_Symbol", "ensembl_id")) %>% 
  distinct()
```

Join symbosl to lm_res
```{r}
lm_res_symbols = 
  lm_res %>% 
  left_join(symbols_list, by = "ensembl_id") %>% 
  relocate(Hugo_Symbol)
```

```{r}
head(lm_res_symbols)
```

Save lm_res
```{r}
saveRDS(lm_res_symbols, file = "/Users/castilln/Desktop/thesis/github/results/differential_expression/transcripts/tcga_non_core_spliceosome_lm_res_symbols.rds")
```
