---
title: "CCLE expression analysis related to spliceosomal mutations"
author: "Leticia Castillon"
date: "23-11-2020"
output:
  html_document:
    df_print: paged
---

```{r, message=FALSE, include=FALSE}
library(tidyverse)
library(edgeR)
library(dplyr)
library(readr)
library(biomaRt)
library(cowplot)
library(ggpubr)
library(ggplot2)
library(ggsci)
library(ggrepel)
library(ggExtra)
library(KEGGREST)
library(hrbrthemes)
library(wesanderson)
library(data.table)
library(Rtsne)
library(RColorBrewer)
library(gridExtra)
select = dplyr::select
rename = dplyr::rename
filter = dplyr::filter
```


source R profile. Memory was set to 500000.
```{r}
Sys.setenv("R_ENVIRON_USER"='/Users/castilln/.Renviron')
Sys.getenv("R_ENVIRON_USER")

```


# Load expression data & spliceosome data
```{r}
transcripts_df <- read_csv("/Users/castilln/Desktop/thesis/localdata/depmap/CCLE_RNAseq_transcripts.csv")
#rows: tumor sample barcode
#column: transcript
#values: expression (TPM) already log(x+1) transformed

#reads - raw counts? 
reads_df <- read_csv("/Users/castilln/Desktop/thesis/localdata/depmap/CCLE_RNAseq_reads.csv")

sample_info <- read_csv("/Users/castilln/Desktop/thesis/localdata/depmap/sample_info.csv") #metadata

CCLE_mutations <- fread("/Users/castilln/Desktop/thesis/localdata/depmap/CCLE_info", header = TRUE) #mutations and sample info 

mutations_spliceosome = fread("/Users/castilln/Desktop/thesis/localdata/depmap/mutations_spliceosome.csv")
```
Contingency table - Annotation of cell lines with spliceosome mutations
```{r}
cell_lines_list = 
  CCLE_mutations %>% 
  select(stripped_cell_line_name, primary_disease, DepMap_ID) 

cell_lines_list_mutated = 
 cell_lines_list %>% 
   mutate(spliceosome_mutated = 
           case_when(
            cell_lines_list$stripped_cell_line_name %in% mutations_spliceosome$stripped_cell_line_name ~ "YES", #if the mutation is in a gene from the spliceosome: 1 
            !cell_lines_list$stripped_cell_line_name %in% mutations_spliceosome$stripped_cell_line_name ~ "NO")) %>% 
  distinct()

cell_lines_list_mutated %>% 
  distinct() %>%
  group_by(primary_disease,spliceosome_mutated) %>%
  tally()

table(distinct(cell_lines_list_mutated)$primary_disease,distinct(cell_lines_list_mutated)$spliceosome_mutated)

spliceosome_mutated_ID = 
  cell_lines_list_mutated %>% 
  select(DepMap_ID, spliceosome_mutated) %>% 
  distinct() 
```


### Differential Expression Analysis
```{r}
#make sure expression data is normally distributed 
#central limit theorem: sample size large enough to assume normal distribution
CCLE_expression = 
  CCLE_expression %>% 
  rename("DepMap_ID" = "V1")

#join expression data with metadata
long_CCLE_expression =
  CCLE_expression %>%
  pivot_longer(cols=-DepMap_ID, names_to = "Gene", values_to = "TPM") %>% 
  left_join(cell_lines_list_mutated, by = "DepMap_ID")

```

#Lets see the distribution of the expression data
```{r}
library(ggdist)
ggplot(long_CCLE_expression, aes(x=TPM)) +
  geom_histogram(aes(y=..density..), position="identity", alpha=0.5) + 
  geom_density(alpha=0.6)
```
Filter out those genes whose median for expression is lower than 1 across all samples
```{r}
median_CCLE_expression =
  long_CCLE_expression %>% 
  group_by(Gene) %>% 
  dplyr::mutate(median = median(TPM, na.rm=TRUE)) %>% 
  filter(median >= 1)
```

Plot again
```{r}
ggplot(median_CCLE_expression, aes(x=TPM)) +
  geom_histogram(aes(y=..density..), position="identity", alpha=0.5) + 
  geom_density(alpha=0.6)
```
After joinin with the expression data, let's see how many cell lines we have mutated or not mutated
```{r}
median_CCLE_expression = 
  median_CCLE_expression %>% 
  rename("cell_line" = "stripped_cell_line_name")

median_CCLE_expression %>% 
  ungroup() %>% 
  select(primary_disease, cell_line, spliceosome_mutated) %>% 
  distinct() %>% 
  group_by(primary_disease, spliceosome_mutated) %>% 
  tally()

median_CCLE_expression %>% 
  ungroup %>% 
  select(cell_line) %>% 
  distinct() %>% dim()
```
Let's keep only those cancers with a relatively significant number of cell lines w/o mutations in the spliceosome: 

```{r}
keep <- c("Bone Cancer", "Pancreatic Cancer", "Leukemia")

df_deseq = 
median_CCLE_expression %>% 
  filter(primary_disease %in% keep)
```

# t-test
```{r}
#filt_CCLE_expression %>% 
  df_deseq %>% 
  select(DepMap_ID, Gene, TPM, spliceosome_mutated) %>% 
  filter(spliceosome_mutated == "NO") %>% 
  select(c(Gene, TPM))-> no_t_expression

#head(no_t_expression)

#filt_CCLE_expression %>% 
df_deseq %>% 
  select(DepMap_ID, Gene, TPM, spliceosome_mutated) %>% 
  filter(spliceosome_mutated == "YES") %>% 
  select(c(Gene, TPM)) -> yes_t_expression

#head(yes_t_expression)

t.test(no_t_expression$TPM, yes_t_expression$TPM, alternative = "two.sided", var.equal = FALSE)
```


# Create DGElist
A DGElist object conveniently stores count matrix, sample metadata and gene annotation as one object for easy manipulation.

## Count matrix
```{r}
df_counts = df_deseq %>% 
  select(cell_line, Gene, TPM) %>%  
  pivot_wider(names_from = cell_line, values_from = TPM) %>%  
  as.data.frame()

#STORE FIRST COLUMN AS ROWNAMES
rownames(df_counts) = df_counts[,1]

rownames(df_counts) <- gsub("\\s*\\([^\\)]+\\)","",as.character(rownames(df_counts)))

#REMOVE FIRST COLUMN
df_counts = df_counts[,-1]  

##THERE ARE NA VALUES IN DF_COUNTS. SUBSTITUTE NA VALUES BY THE MEAN EXPRESSION FOR THAT CELL LINE
for(i in 1:ncol(df_counts)){
  df_counts[is.na(df_counts[,i]), i] <- mean(df_counts[,i], na.rm = TRUE)
} 
```

```{r}
meta = 
  df_deseq %>% 
  ungroup() %>% 
  select(-c("TPM", "median", "Gene")) %>% 
  distinct()
```

Reorder count columns by meta ID

```{r}
##REORDER COLUMNS BY SAMPLE INFO
df_counts = df_counts[meta$cell_line]

head(df_counts)
```
Create grouping factor

```{r}
meta = 
  meta %>% 
  mutate(group = as.factor(spliceosome_mutated)) %>% 
  select(-spliceosome_mutated) %>% 
  mutate(primary_disease = as.factor(primary_disease))
```
 

Create gene annotation 

```{r}
#SOMETIMES PROBLEMS TO LOAD THIS LIBRARIES BECAUSE INCOMPATIBILITIES W/ COMPLEX HEATMAP: START NEW R SESSION
library("AnnotationDbi")
library("org.Hs.eg.db")

gene_annot = AnnotationDbi::mapIds(org.Hs.eg.db, keys = rownames(df_counts),
                                  column = c("GENENAME"), keytype = ("SYMBOL")) %>% 
  as.data.frame()

colnames(gene_annot) = c("DESCRIPTION")
```

## Make grouping factor in DGEList
Change primary_disease names to something understandable for R as names

```{r}
meta$primary_disease = gsub(" ", "_", meta$primary_disease)
meta$primary_disease = gsub("/", "_", meta$primary_disease)
```


```{r}
dge = DGEList(counts = df_counts, 
              samples = meta, 
              genes = gene_annot)

summary(dge)
```

# Filtering and pre-process
Filter low expressed genes

```{r}
#keep  = filterByExpr(dge, group = group)
#dge = dge[keep, , keep.lib.sizes = FALSE]
```

Calculate library normalization factors

```{r}
dge <- calcNormFactors(dge)
dge$samples
```
# MDS (multidimensional scaling plot)
```{r}
plotMDS(dge)
```




# Differential expression
Set up desing matrix

```{r}
design = model.matrix(~0+group+primary_disease, data = dge$samples)


colnames(design)
#design
```

Calculate dispersions

```{r}
dge = estimateDisp(dge, design)
```

```{r}
fit <- glmQLFit(dge, design)

plotQLDisp(fit)
```

In order to compare the difference in cell lines with spliceosome mutations vs wt for all different cancers, we need to define some contrasts for the model. 

```{r}
my.contrasts <- makeContrasts(
  YvsN = `groupYES`-`groupNO`,
  levels = design
)
```

Now we are going to use the model contrasts to determine pairwise differential expression using the quasi-likelihood (QL)-F-test

```{r}
qlf.spliceosome.YvsN <- glmQLFTest(fit, contrast=my.contrasts[,"YvsN"])
topTags(qlf.spliceosome.YvsN)
```


Adjust p-values (FDR) and merge data into a single table to facilitate the analysis: 

```{r}
TidyQLF = function(qlf, contrast) {
  df_qlf =
    qlf$table %>%
    as_tibble(rownames = "SYMBOL") %>%
    mutate(FDR = p.adjust(PValue, method = "fdr"))
  
  df_full =
    cbind(qlf$genes,df_qlf) %>% 
    mutate(comparison = contrast) %>%
    relocate(SYMBOL, .before = DESCRIPTION) %>% 
    relocate(comparison, .before = SYMBOL) 
  
  return(df_full)
}


tbl_YvsN = TidyQLF(qlf.spliceosome.YvsN,"YesvsNo")


head(tbl_YvsN)
```


## Exploration using PCA/tSNE

Prepare for tSNE/PCA

```{r}
##PIVOT WIDER
df_wide = 
  median_CCLE_expression %>% 
  select(DepMap_ID, Gene, TPM, spliceosome_mutated, primary_disease) %>% 
  pivot_wider(names_from = Gene, values_from = TPM) %>%  
  as.data.frame() %>% 
  relocate(DepMap_ID, .after = primary_disease)



expression_pca <- as.matrix(df_wide[,3:11400])
expression <- as.matrix(df_wide[,4:11400])
disease <- df_wide[, 2] #disease
mutated <- df_wide[, 1]

##GET RID OF NA VALUES AND SUBSTITUTE FOR MEAN EXPRESSION PER CELL LINE
for(i in 1:ncol(expression)){
  expression[is.na(expression[,i]), i] <- mean(expression[,i], na.rm = TRUE)
} 
```

Run PCA

```{r,  fig.height=4, fig.width=7}
pca1 = prcomp(expression, center = TRUE, scale = TRUE, na.action=na.omit(expression))

plotData = pca1$x[,1:2]
plotData = cbind(DepMap_ID = expression_pca[,1], plotData)
rownames(plotData) = NULL

head(plotData)

ID = plotData[,1]

to_join = 
  median_CCLE_expression %>% 
  select(DepMap_ID, primary_disease, spliceosome_mutated)

plotData %>% 
  as_tibble() %>% 
  left_join(to_join, by ="DepMap_ID") %>%
  mutate(PC1 = as.double(PC1),
         PC2 = as.double(PC2))  -> plotData_gg
```
Plot

```{r,  fig.height=4, fig.width=12}

ggplot(plotData_gg, aes(x = PC1,y = PC2, color = spliceosome_mutated)) +
  geom_point() +
  facet_wrap(vars(primary_disease), scales = "free") + 
  stat_ellipse() -> gg_expression_mutated

ggsave("/Users/castilln/Desktop/thesis/gg_expression_mutated.png", device = "png", width = 35, height = 20, units = "cm", dpi = "retina")
  

ggplot(plotData_gg, aes(x = PC1,y = PC2, color = primary_disease)) +
  geom_point() +
  scale_color_manual(values = c("gainsboro", 'forestgreen', 'red2', 'orange',  'cornflowerblue', 
                'magenta', 'darkolivegreen4',  'indianred1',  'tan4', 'darkblue', 
                'mediumorchid1', 'firebrick4',  'yellowgreen', 'lightsalmon', 'tan3',
                "tan1",  'darkgray','wheat4',  '#DDAD4B',  'chartreuse', 
                 'seagreen1', 'moccasin',   'mediumvioletred', 'seagreen', 'cadetblue1',
                "darkolivegreen1" ,"tan2" ,  "tomato3" , "#7CE3D8", "black", "yellow", "violetred", "blue")) -> gg_expression_disease

  ggsave("/Users/castilln/Desktop/thesis/gg_expression_disease.png", device = "png", width = 25, height = 15, units = "cm", dpi = "retina")

grid.arrange(gg_expression_disease, gg_expression_mutated, ncol = 2)
  
```


Run tSNE

```{r, fig.height=5, fig.width=5}
set.seed(6)
tsne_results <- Rtsne(expression, dims = 2, perplexity = 30, theta = 0.5, check_duplicates = FALSE, pca = TRUE)

#plot

plotData =
  tsne_results$Y %>% 
  as_tibble() %>% 
  mutate(DepMap_ID = ID)
 
plotData = 
  plotData %>% 
  left_join(to_join, by ="DepMap_ID") %>% 
  mutate(V1 = as.double(V1),
         V2 = as.double(V2))
 
plotData %>% 
  ggplot(aes(x = V1,y = V2, color = spliceosome_mutated)) +
  geom_point(aes(alpha=0.5), size = 1) +
  xlab(element_blank()) +
  ylab(element_blank()) +
  facet_wrap(vars(primary_disease)) +
  theme_bw() +
  ggtitle("tSNE") +
  theme(legend.position = "bottom",
        aspect.ratio = 1,
        axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks = element_blank(),
        plot.title = element_text(face = "bold")) +
  scale_color_manual(values = c('forestgreen', 'red2')) ->tsne_expression

ggsave("/Users/castilln/Desktop/thesis/tsne_expression.png", device = "png", width = 30, height = 25, units = "cm", dpi = "retina")

print(tsne_expression)
```





