---
title: "Raw reads expression analysis"
output: html_notebook
---


```{r, message=FALSE, include=FALSE}
library(tidyverse)
library(edgeR)
library(readr)
library(biomaRt)
library(data.table)
select = dplyr::select
rename = dplyr::rename
filter = dplyr::filter
```

source R profile. Memory was set to 500000.
```{r}
Sys.setenv("R_ENVIRON_USER"='/Users/castilln/.Renviron')
Sys.getenv("R_ENVIRON_USER")

```

# Load expression data & spliceosome data
```{r}
raw_reads <- read_csv("/Users/castilln/Desktop/thesis/localdata/depmap/CCLE_RNAseq_reads.csv") 
#rows: tumor sample barcode
#column: genes

sample_info <- read_csv("/Users/castilln/Desktop/thesis/localdata/depmap/sample_info.csv") #metadata

CCLE_mutations <- fread("/Users/castilln/Desktop/thesis/localdata/depmap/CCLE_info", header = TRUE) #mutations and sample info 

mutations_spliceosome = fread("/Users/castilln/Desktop/thesis/localdata/depmap/mutations_spliceosome.csv")
```

Contingency table - Annotation of cell lines with spliceosome mutations
```{r}
cell_lines_list = 
  CCLE_mutations %>% 
  select(stripped_cell_line_name, primary_disease, DepMap_ID) 

cell_lines_list_mutated = 
 cell_lines_list %>% 
   mutate(spliceosome_mutated = 
           case_when(
            cell_lines_list$stripped_cell_line_name %in% mutations_spliceosome$stripped_cell_line_name ~ "YES", #if the mutation is in a gene from the spliceosome: 1 
            !cell_lines_list$stripped_cell_line_name %in% mutations_spliceosome$stripped_cell_line_name ~ "NO")) %>% 
  distinct()

cell_lines_list_mutated %>% 
  distinct() %>%
  group_by(primary_disease,spliceosome_mutated) %>%
  tally()

table(distinct(cell_lines_list_mutated)$primary_disease,distinct(cell_lines_list_mutated)$spliceosome_mutated)

spliceosome_mutated_ID = 
  cell_lines_list_mutated %>% 
  select(DepMap_ID, spliceosome_mutated) %>% 
  distinct() 
```

```{r}
#Rename so that raw reads table has same variable name as metadata
raw_reads = 
  raw_reads %>% 
  rename("DepMap_ID" = "X1")

#join expression data with metadata
longraw_reads_info =
  raw_reads %>%
  pivot_longer(cols=-DepMap_ID, names_to = "Gene", values_to = "counts") %>% 
  left_join(cell_lines_list_mutated, by = "DepMap_ID")

head(longraw_reads_info)
```
Let's keep only those cancers with a relatively significant number of cell lines w/o mutations in the spliceosome:
```{r}
keep <- c("Bone Cancer", "Breast Cancer", "Fibroblast", "Leukemia", "Lung Cancer", "Ovarian Cancer", "Pancreatic Cancer", "Sarcoma", "Skin Cancer")

df_deseq = 
longraw_reads_info %>% 
  filter(primary_disease %in% keep)
```

Create a new column for ensembl gene ID 
```{r}
df_deseq = 
  df_deseq %>%
    separate(col = "Gene", into = c("Hugo_Symbol", "ensembl_id"), sep = "\\s") 
   
df_deseq$ensembl_id <- gsub("\\(|\\)","",as.character(df_deseq$ensembl_id))

head(df_deseq)
```


## Preprocessing
Let's visualize the distribution of the counts
```{r}
library(ggdist)
ggplot(df_deseq, aes(x=counts)) +
  geom_histogram(aes(y=..density..), position="identity", alpha=0.5) + 
  geom_density(alpha=0.6)
```

Filter out those genes whose median for expression is lower than 1 across all samples
```{r}
median_counts =
  df_deseq %>% 
  group_by(Hugo_Symbol) %>% 
  dplyr::mutate(median = median(counts, na.rm=TRUE)) %>% 
  filter(median >= 1)
```

Plot again
```{r}
ggplot(median_counts, aes(x=counts)) +
  geom_histogram(aes(y=..density..), position="identity", alpha=0.5) + 
  geom_density(alpha=0.6)
```

Let's calculate the counts-per-million (CPM) for each gene 
```{r}
df_deseq = df_deseq %>%
  group_by(DepMap_ID) %>%
  mutate(cpm = 1e6 * (counts / sum(counts))) %>%
  ungroup()
```

# DGE List
A DGElist is an object that stores count matrix, sample metadata and gene annotation as one object and is used by edgeR

Create count matrix
```{r}
df_counts = df_deseq %>% 
  select(ensembl_id,DepMap_ID,counts) %>% 
  pivot_wider(names_from = DepMap_ID ,values_from = counts) %>% 
  as.data.frame()

#REMOVE NA VALUES FROM ENSEMBL ID COLUMN
df_counts <- df_counts[!is.na(df_counts$ensembl_id),]

#STORE FIRST COLUMN AS ROWNAMES
rownames(df_counts) = df_counts[,1]

#REMOVE FIRST COLUMN
df_counts = df_counts[,-1]  

#CHECK IF THERE ARE NA VALUES
sum(is.na(df_counts))

##THERE ARE NA VALUES IN DF_COUNTS. SUBSTITUTE NA VALUES BY THE MEAN EXPRESSION FOR THAT CELL LINE
#for(i in 1:ncol(df_counts)){
# df_counts[is.na(df_counts[,i]), i] <- mean(df_counts[,i], na.rm = TRUE)
#} 
```

Create meta object. Meta only includes the data not specified in the raw reads file, therefore only cell line, spliceosome mutated status, primary disease and DepMap ID, which is what we are using as identifier. 
```{r}
meta = 
  df_deseq %>% 
  ungroup() %>% 
  select(-c("counts", "Hugo_Symbol", "ensembl_id", "cpm")) %>% 
  distinct()

#CHANGE PRIMARY DISEASE NAMES SO THAT R CAN UNDERSTAND THEM, NECESSARY FOR DOWNSTREAM ANALYSIS
meta$primary_disease = gsub(" ", "_", meta$primary_disease)
meta$primary_disease = gsub("/", "_", meta$primary_disease)

head(meta)
```

Create grouping factor
```{r}
meta = 
  meta %>% 
  mutate(group = as.factor(spliceosome_mutated)) %>% 
  select(-spliceosome_mutated) %>% 
  mutate(primary_disease = as.factor(primary_disease))
```


Reorder count columns by meta ID
```{r}
##REORDER COLUMNS BY SAMPLE INFO
df_counts = df_counts[meta$DepMap_ID]

##NEED TO BE CHARACTER SO THAT THEY CAN BE STORED IN DGEList
df_counts[] <- lapply(df_counts, function(x) as.numeric(as.character(x)))

head(df_counts)

```

Create gene annotation
```{r}
#SOMETIMES PROBLEMS TO LOAD THIS LIBRARIES BECAUSE INCOMPATIBILITIES W/ COMPLEX HEATMAP: START NEW R SESSION
library("AnnotationDbi")
library("org.Hs.eg.db")


gene_annot = AnnotationDbi::mapIds(org.Hs.eg.db, keys=rownames(df_counts),
       column=c("SYMBOL"), keytype="ENSEMBL") %>% 
  as.data.frame()

gene_desc = AnnotationDbi::mapIds(org.Hs.eg.db, keys=rownames(df_counts),
       column=c("GENENAME"), keytype="ENSEMBL") %>% 
  as.data.frame()

gene_annot = cbind(gene_annot,gene_desc)

colnames(gene_annot) = c("SYMBOL","DESCRIPTION")

head(gene_annot)
```


Now, let's create the DGElist
```{r}
dge = DGEList(counts = df_counts,
              samples = meta,
              genes = gene_annot)

summary(dge)
```

# Filtering and pre-process

Remove genes for which we do not know annotation: NA Symbol. In order to avoid duplicates
```{r}
i <- is.na(dge$genes$SYMBOL)
dge <- dge[!i, ]
```

Filter lowly expressed genes
```{r}
#keep = filterByExpr(dge)
#dge = dge[keep, , keep.lib.sizes=FALSE]
```

Calculate library normalization factors (effective library size)
```{r}
#dge <- calcNormFactors(dge)
#dge$samples
```

# MDS
Plot MDS (quality control to detect outliers/bad samples)
```{r}
plotMDS(dge)
```


# Differential expression
Set up design matrix
```{r}
design = model.matrix(~primary_disease + group, data = dge$samples)

colnames(design)
#design
```


Filter lowly expressed genes (here they suggets to do it after setting up a design: https://www.biostars.org/p/430379/) 
```{r}
#keep = filterByExpr(dge, design)

#IF WE DO NOT WANT TO TEST FOR DIFFERENTIAL EXPRESSION BETWEEN CANCERS:
keep = filterByExpr(dge, group = dge$sample$group)

dge = dge[keep, , keep.lib.sizes=FALSE]
```

Calculate library normalization factors (effective library size)
```{r}
dge <- calcNormFactors(dge)
dge$samples
```

Estimate dispersion
```{r}
dge = estimateDisp(dge, design)
```

Fit a negative binomial generalized linear model to our data
```{r}
fit <- glmQLFit(dge, design)

plotQLDisp(fit)
```

In order to compare the difference in cell lines with spliceosome mutations vs wt for all different cancers, we need to define some contrasts for the model. 
```{r}
my.contrasts <- makeContrasts(
  YvsN = `groupYES`- `groupNO`,
  levels = design
)
```

Now we are going to use the model contrasts to determine pairwise differential expression using the quasi-likelihood (QL)-F-test. 
As we define glmQLFTest, we are detecting the genes that are differentially expressed in response to the spliceosome machinery, adjusting for baseline differences between cancers.  
```{r}
qlf.spliceosome.YvsN <- glmQLFTest(fit)
topTags(qlf.spliceosome.YvsN)
```
Likelihood radio tests
```{r}
lrt <- glmLRT(fit)
topTags(lrt)
```
```{r}
summary(decideTests(qlf.spliceosome.YvsN))
```

------------------------------------
Duplicate elements?  FIX
```{r}
#THERE SEEM TO BE SOME DUPLICATED ELEMENTS. REMOVE NA
qlf.spliceosome.YvsN %>% 
  duplicated()

view(qlf.spliceosome.YvsN)
```

Adjust p-values (FDR) and merge data into a single table to facilitate the analysis: 
```{r}
TidyQLF = function(qlf, contrast) {
  df_qlf =
    qlf$table %>%
    as_tibble(rownames = "SYMBOL") %>%
    mutate(FDR = p.adjust(PValue, method = "fdr"))
  
  df_full =
    cbind(qlf$genes,df_qlf) %>% 
    mutate(comparison = contrast) %>%
    relocate(SYMBOL, .before = DESCRIPTION) %>% 
    relocate(comparison, .before = SYMBOL) 
  
  return(df_full)
}


tbl_YvsN = TidyQLF(qlf.spliceosome.YvsN,"YesvsNo")


head(tbl_YvsN)
```