---
title: "MSK-IMPACT"
output: html_notebook
---


```{r}
library(tidyverse)
library(data.table)
library(biomaRt)
```

Set wd 
```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = "/Users/castilln/Desktop/thesis/localdata")
```

```{r}
library(readr)
data_mutations_mskcc <- read_delim("msk-impact/msk_impact_2017/data_mutations_mskcc.txt", 
    "\t", escape_double = FALSE, trim_ws = TRUE, 
    skip = 1)

drug_rep <- fread("depmap/repurposing_drugs_20200324.txt")

ccle <- fread("depmap/CCLE_info")
```

##FIND OUT EXON LOCATIONS FOR EACH TRANSCRIPT
```{r}
#RETRIEVE TRANSCRIPT LIST A DF
transcript = 
  ccle %>% 
  pull(Annotation_Transcript) %>% 
  as.data.frame() %>% 
  unique()

##REMOVE VERSION AFTER TRANSCRIPT ID
transcript_sub <- map_df(transcript, ~ gsub("\\..*", "", .x)) 

transcript_list <- as.list(transcript_sub)
```

Use biomart to get exon position
```{r}
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl") 

attributes.1 = c("exon_chrom_start", "exon_chrom_end", "transcript_length", "ensembl_gene_id","ensembl_transcript_id","ensembl_exon_id", "chromosome_name", "strand")
transcript_data<-getBM(attributes.1, filters = c('ensembl_transcript_id'), values=transcript_list, mart=ensembl)
```
Clean dfs
```{r}
mskcc = 
  data_mutations_mskcc %>% 
  dplyr::select(-c("Entrez_Gene_Id", "Center")) %>% 
  dplyr::rename("SYMBOL" = "Hugo_Symbol")
```

# Prepare drug df
Keep only oncology drugs
```{r}
drug_onc = 
  drug_rep %>% 
  filter(disease_area == "oncology") %>% 
  drop_na()

head(drug_onc)
```

Parse multiple target Symbols to different columns
```{r}
library(splitstackshape)
pars_drug_onc = cSplit(as.data.table(drug_onc), "target", "|")

pars_drug_onc = 
  pars_drug_onc %>% 
  dplyr::select(-c("clinical_phase", "disease_area"))

head(pars_drug_onc)

#PIVOT LONGER
target_list = 
  pars_drug_onc %>% 
  pivot_longer(cols = -c("pert_iname", "moa", "indication"), names_to = "t", values_to = "SYMBOL") %>% 
  dplyr::select(-t) %>% 
  drop_na()

head(target_list)
```


# Join with MSK-IMPACT list 
```{r}
head(mskcc)
```

```{r}
mskcc_drug = 
  mskcc %>% 
  left_join(target_list, by = "SYMBOL") %>% 
  unique()

head(mskcc_drug)
```
Keep only those actionable by drugs
```{r}
mskcc_act = 
  mskcc_drug %>% 
  drop_na(pert_iname)

##SELECT FEATURES OF INTEREST TO JOIN WITH THE OUTPUT FROM SPLICEAI
mskcc_sel = 
  mskcc_act %>% 
  dplyr::select(c("SYMBOL", "Consequence", "Transcript_ID", "pert_iname", "moa", "indication", "Variant_Type", "Chromosome", "Strand"))

head(mskcc_act)
```

# Load spliceAI output
```{r}
##OBS THIS IS THE OUTPUT FILTERED FOR THOSE CLASSIFIED AS SPLICED_ALTERED VARIANT > 0.5 PROBABILITY
splice_out = readRDS("../github/results/spliceAI/ths5.rds")

splice_out = 
  splice_out %>% 
  dplyr::rename("Transcript_ID" = "Annotation_Transcript")

splice_out$Transcript_ID <- gsub("\\..*", "", splice_out$Transcript_ID)

head(splice_out)
```
Join spliceAI output with list mskcc-drug
```{r}
join_df = 
  splice_out %>% 
  left_join(mskcc_sel, by = "SYMBOL") %>% 
  drop_na(pert_iname) %>% 
  unique()

symbol_join_list = 
  join_df %>% 
  pull(SYMBOL) %>% 
  unique()
```

```{r}
head(join_df)
```




Map variants to exon/intron
```{r}
test_df = 
  join_df %>% 
  dplyr::select(c("DepMap_ID", "POS", "SYMBOL", "SCORE_ACC_GAIN", "SCORE_ACC_LOSS", "SCORE_DONOR_GAIN", "SCORE_DONOR_LOSS","POS_ACC_GAIN", "POS_ACC_LOSS", "POS_DONOR_GAIN", "POS_DONOR_LOSS", "Transcript_ID.x", "Chromosome", "Strand"))

test_df = 
  test_df %>% 
  dplyr::rename("Transcript_ID" = "Transcript_ID.x")

transcript_data = 
  transcript_data %>% 
  dplyr::rename("Transcript_ID" = "ensembl_transcript_id")
```

```{r}
head(transcript_data)
```


```{r}
test_transcript = 
  test_df %>% 
  left_join(transcript_data, by = "Transcript_ID")

test_exon = 
  test_transcript %>% 
  dplyr::select(c("SYMBOL", "POS", "Transcript_ID", "exon_chrom_start", "exon_chrom_end", "transcript_length"))

head(test_transcript)

for (i in 1:length(test_transcript){
  test_exon[[i]] %>% 
  mutate(MAX_PROBABILITY = apply(X = data[,4:7], MARGIN = 1, FUN = max)
         }
```

```{r}
head(join_df)
```
Create new matrix only containing positions with probability > 0.5 of generating a new acceptor position
```{r}
max_acc = 
  join_df %>% 
  dplyr::select(-"MAX_PROBABILITY") %>% 
  filter(SCORE_ACC_GAIN > 0.5) 

max_acc$POS_ACC_GAIN <- as.numeric(max_acc$POS_ACC_GAIN)


head(max_acc)
```
Generate new position in the bp where mutation occurs
```{r}
max_acc_mut = 
  max_acc %>% 
  mutate(MUT_POS = POS + POS_ACC_GAIN)

head(max_acc_mut)
```

Create new matrix only containing positions with probability > 0.5 of generating a new donor position
```{r}
max_don = 
  join_df %>% 
  dplyr::select(-"MAX_PROBABILITY") %>% 
  filter(SCORE_DONOR_GAIN > 0.5) 

max_don$POS_DONOR_GAIN <- as.numeric(max_don$POS_DONOR_GAIN)


head(max_don)
```
Generate new position in the bp where mutation occurs
```{r}
max_don_mut = 
  max_don %>% 
  mutate(MUT_POS = POS + POS_DONOR_GAIN)

head(max_don_mut)
```

Merge both data frames 
```{r}
max_mut <- rbind(max_acc_mut, max_don_mut)

head(max_mut)
```

CAREFULT THAT WE ARE NOT CONSIDERING MUTATIONS THAT OCCUR ONLY BECAUSE OF LOSS OF SPLICING

Add Hugo Symbol to transcript and exon length data
```{r}
##REMOVE VERSION AFTER TRANSCRIPT ID
ccle$Annotation_Transcript <- gsub("\\..*", "", ccle$Annotation_Transcript)

ccle_ann = 
  ccle %>% 
  dplyr::select(c("Hugo_Symbol", "Annotation_Transcript")) %>% 
  dplyr::rename("SYMBOL" = "Hugo_Symbol", 
                "Transcript_ID" = "Annotation_Transcript")


##JOIN SYMBOL TO TRANSCRIPT DATA
transcript_symbol = 
  transcript_data %>% 
  dplyr::rename("Transcript_ID" = "ensembl_transcript_id") %>% 
  left_join(ccle_ann, by = "Transcript_ID")
```

Genomic ranges for exon length data
```{r}
library(GenomicRanges)
##GENERATE GR OBJECT FROM BIOMART DATA - GET EXON INTERVALS
bm_intervals = GRanges(seqnames = transcript_symbol$SYMBOL, ranges = IRanges(start = transcript_symbol$exon_chrom_start, end = transcript_symbol$exon_chrom_end),
               strand = transcript_symbol$strand)
```

Prepare spliceAI output modified for mutation position for GenomicRanges
```{r}
head(max_mut)
```

```{r}
max_mut_ranges = 
  max_mut %>% 
```

