---
title: "Shannon entropy"
output: html_notebook
---

Load libraries
```{r}
library(data.table)
library(tidyverse)
library(EntropyExplorer)
```

Set wd 
```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = "/Users/castilln/Desktop/thesis/localdata")
```
source R profile. Memory was set to 500000.
```{r}
Sys.setenv("R_ENVIRON_USER"='/Users/castilln/.Renviron')
Sys.getenv("R_ENVIRON_USER")

```
Load data
```{r}
transcripts = read_csv("depmap/CCLE_RNAseq_transcripts.csv")

#METADATA
sample_info <- read_csv("depmap/sample_info.csv") #metadata

#MUTATIONS AND SAMPLE INFO
CCLE_mutations <- fread("depmap/CCLE_info", header = TRUE) 

#LIST OF MUTATIONS FROM CCLE MUTATIONS THAT AFFECT THE SPLICEOSOME
mutations_spliceosome = fread("depmap/mutations_spliceosome.csv")
```
Contingency table - Annotation of cell lines with spliceosome mutations
```{r}
cell_lines_list = 
  CCLE_mutations %>% 
  select(stripped_cell_line_name, primary_disease, DepMap_ID) 

cell_lines_list_mutated = 
 cell_lines_list %>% 
   mutate(spliceosome_mutated = 
           case_when(
            cell_lines_list$stripped_cell_line_name %in% mutations_spliceosome$stripped_cell_line_name ~ "YES", #if the mutation is in a gene from the spliceosome: 1 
            !cell_lines_list$stripped_cell_line_name %in% mutations_spliceosome$stripped_cell_line_name ~ "NO")) %>% 
  distinct()

cell_lines_list_mutated %>% 
  distinct() %>%
  group_by(primary_disease,spliceosome_mutated) %>%
  tally()

table(distinct(cell_lines_list_mutated)$primary_disease,distinct(cell_lines_list_mutated)$spliceosome_mutated)

spliceosome_mutated_ID = 
  cell_lines_list_mutated %>% 
  select(DepMap_ID, spliceosome_mutated) %>% 
  distinct() 
```

Pivot to long 
```{r}
#RENAME FIRST VARIABLE 
transcripts = 
  transcripts %>% 
  rename("DepMap_ID" = "X1")

#JOIN TRANSCRIPT INFO WITH METADATA
long_transcripts_info = 
  transcripts %>% 
  pivot_longer(cols=-DepMap_ID, names_to = "Gene", values_to = "log_TPM") %>% 
  left_join(cell_lines_list_mutated, by = "DepMap_ID")

head(long_transcripts_info)
```

Let's keep only those cancers with a relatively significant number of cell lines w/o mutations in the spliceosome:
```{r}
keep <- c("Bone Cancer", "Fibroblast", "Leukemia", "Lung Cancer", "Ovarian Cancer", "Pancreatic Cancer", "Sarcoma", "Skin Cancer")

df_transcripts = 
long_transcripts_info %>% 
  filter(primary_disease %in% keep)

head(df_transcripts)
```
filter
```{r}
no_transcripts = 
  df_transcripts %>% 
  filter(spliceosome_mutated == "NO")

yes_transcripts = 
  df_transcripts %>% 
  filter(spliceosome_mutated == "YES")
```

entropy calculation
pre-process: both matrices need to have same number of rows; transcripts as rows
```{r}
no_transcripts = 
  no_transcripts %>% 
  select(c("DepMap_ID", "Gene", "log_TPM")) %>% 
  pivot_wider(id_cols = "Gene", names_from = "DepMap_ID", values_from = "log_TPM")
```

```{r}
yes_transcripts = 
  yes_transcripts %>% 
  select(c("DepMap_ID", "Gene", "log_TPM")) %>% 
  pivot_wider(id_cols = "Gene", names_from = "DepMap_ID", values_from = "log_TPM")
```

fix rownames
```{r}
yes_transcripts = as.data.frame(yes_transcripts)

rownames(yes_transcripts) <- yes_transcripts[,1]

yes_transcripts = yes_transcripts[,-1]

yes_transcripts[is.na(yes_transcripts)] <- 0
```

```{r}
no_transcripts = as.data.frame(no_transcripts)

rownames(no_transcripts) <- no_transcripts[,1]

no_transcripts = no_transcripts[,-1]
```


```{r}
head(yes_transcripts)
sum(yes_transcripts < 0)
```
Add +1 to each value in df (necessary to calculate the entropy)
```{r}
yes_transcripts <- yes_transcripts + 1
no_transcripts <- no_transcripts +1
```

```{r}
head(yes_transcripts)
```

Calculate value of Shannon Entropy for both groups
```{r} 
shannon = EntropyExplorer(yes_transcripts, no_transcripts, "dse", "v")   
```

```{r}
shannon_df = 
  shannon %>% 
  as.data.frame()

head(shannon_df)
```

Save as RDS object
```{r}
#saveRDS(shannon_df, "../../localdata/shannon/res_shannonEntropy.rds")
```

```{r}
names <- rownames(shannon_df) 
rownames(shannon_df) <- NULL
shannon_df = cbind(names, shannon_df)

shannon_df = 
  shannon_df %>% 
  rename("SYMBOL" = "names")

head(shannon_df)
```
Separate SYMBOL column into SYMBOL and entrez id 
```{r}
shannon_df = 
  shannon_df %>%
    separate(col = "SYMBOL", into = c("SYMBOL", "ENTREZ"), sep = "\\s") 

##Note that some did not have a symnbol id
   
shannon_df$ENTREZ <- gsub("\\(|\\)","",as.character(shannon_df$ENTREZ))

head(shannon_df)
```

```{r}
df_transcripts = 
  df_transcripts %>% 
  select(c("DepMap_ID", "Gene", "log_TPM"))

head(df_transcripts)
```

```{r}
library(TFBSTools)

##CALCULATE ENTROPY PER GENE PER CELL LINE
##LIST OF CANCERS
list_cell_lines = 
  df_transcripts %>% 
  pull(DepMap_ID) %>% 
  unique()

##EMPTY LIST TO STORE LINEAR MODELS
shannon_list <- list()

for (dis in list_cell_lines){
  dis_df = 
    df_transcripts %>% 
    filter(DepMap_ID == dis)
  
  
  ##LINEAR MODEL
  shannon_list[[dis]] = 
  dis_df %>% 
  select(log_TPM) %>% 
  shannon.entropy()
}
```

```{r}
saveRDS(shannon_list, file = "../results/shannon_list_cell_line.rds")
```

T-test
```{r}
library(broom)
shannon_yes = shannon_df$`SE(expm1)`
shannon_no = shannon_df$`SE(expm2)`

wilc_res = t.test(shannon_yes, shannon_no)

tidy(wilc_res)
```


