---
title: "Gene effect off-target annotation - MSK IMPACT"
output: html_notebook
---
Libraries
```{r}
library(tidyverse)
library(data.table)
library(ggbeeswarm)
library(readr)
library(ggpubr)


select = dplyr::select
rename = dplyr::rename
```

source R profile. Memory was set to 500000.
```{r}
Sys.setenv("R_ENVIRON_USER"='/Users/castilln/.Renviron')
Sys.getenv("R_ENVIRON_USER")

```

Set wd 
```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = "/Users/castilln/Desktop/thesis/localdata")
```

Load data
```{r}
#DEPMAP somatic mutations
ccle = fread("depmap/ccle_info_21q1.csv")
ccle = 
  ccle %>% 
  rename("SYMBOL" = "Hugo_Symbol")

#CRISPR gene effect
gene_effect = fread("depmap/achilles/Achilles_gene_effect.csv")

#LOAD ACTIONABLE SPLICE OUTPUT (ALREADY FILTERED FOR MSK) 
actionable_splice = readRDS("msk-impact/spliceOut_actionable.rds")
head(actionable_splice)
```

Extract MSK IMPACT genes present in the spliceAI output
(This data frame has already been filtered for MSK IMPACT genes)
```{r}
actionable_genes =
  actionable_splice %>% 
  select(SYMBOL) %>% 
  distinct() %>% 
  pull()

head(actionable_genes)
```

Select variables of interest from ccle and filter out MSK IMPACT genes
```{r}
df_mutation_meta = 
  ccle %>% 
  select(DepMap_ID, SYMBOL, primary_disease, Variant_Type, isDeleterious, Variant_Classification) %>% 
  filter(SYMBOL %in% actionable_genes)
head(df_mutation_meta)
```

Pivot long effect matrix and filter for MSK IMPACT genes
```{r}
##PIVOT LONG EFFECT MATRIX
df_crispr = 
  gene_effect %>% 
  pivot_longer(cols = -DepMap_ID, names_to = "SYMBOL", values_to = "gene_effect") %>% 
  mutate(SYMBOL = gsub(" .*","",SYMBOL)) %>% 
  filter(SYMBOL %in% actionable_genes)

head(df_crispr)
dim(df_crispr)
```

Add splice variant annotation to spliceAI output data frame
Since this data has already been filtered for SNPs producing a splice variant with >0.5 confidence, as predicted by the NN, we can simply tag all SNPs as TRUE
```{r, fig.height=10, fig.width=10}
df_splice = 
  actionable_splice %>% 
  mutate(splice_variant = TRUE) %>%
  select(DepMap_ID,SYMBOL,primary_disease,splice_variant,MAX_PROBABILITY, exon_intron)

head(df_splice)
dim(df_splice)
```

Filter out SNPs without cell line annotation (this is due to different versions of CCLE data being used)
```{r}
df_splice = 
  df_splice %>% 
  filter(!is.na(DepMap_ID))
dim(df_splice)
```

Create new column with custom isDeleterious information. isDeleterious only if mutation is an insertion or deletion
```{r}
df_mutation_delet = 
  df_mutation_meta %>% 
  mutate(isDel = ifelse(Variant_Type == "SNP", FALSE, TRUE)) %>% 
  select(-isDeleterious)

dim(df_mutation_delet)
head(df_mutation_delet)
```

Now, let's join the spliceAI output with the CCLE mutation annotation information.
```{r}
df_splice_annot = 
  df_splice %>% 
  left_join(df_mutation_delet, by = c("SYMBOL", "DepMap_ID", "primary_disease")) %>% 
  distinct() %>% 
  #indicate those splice variants that occur in a different gene than the one we gave as input
  mutate(off_target = ifelse(is.na(Variant_Type), TRUE, FALSE)) %>% 
  #Substitute NAs from Variant_Type with "SNP"
  mutate(Variant_Type = ifelse(is.na(Variant_Type), "SNP", Variant_Type)) 

#Create new dataframe exclusively for offtarget mutations
df_off_target = 
  df_splice_annot %>% 
  filter(off_target == T) %>% 
  select(-c("Variant_Classification", "isDel"))

view(df_off_target)

#write_csv(df_off_target, "msk-impact/off_targe_spliceAIout_actionable.csv")

dim(df_splice_annot)
head(df_splice_annot)
```