---
title: "Parsing SpliceAI output"
author: "Leticia Castillon"
date: "27-11-2020"
output: html_notebook
---


Libraries
```{r, message=FALSE, include=FALSE}
library(tidyverse)
library(readr)
library(biomaRt)
library(data.table)
library(readxl)
library(stringr)
library(purrr)
library(sjmisc)
select = dplyr::select
rename = dplyr::rename
```


## Load output from spliceAI
# Read .vcf data
Load tests
```{r}
test2 <- fread("./test2.csv")
```

Read data and merge results for each chromosome in a single file
```{r, message=FALSE}
data_files <- list.files("/Users/castilln/Desktop/thesis/localdata/spliceai/output/")

#read all .vcf files
splice_output_list = list()
for (i in 1:21) {
  splice_output_list[[i]] <- assign(paste0("splice_chr", i),
      fread(paste0("/Users/castilln/Desktop/thesis/localdata/spliceai/output/", data_files[i]), header = TRUE, sep = "\t"))
}

#bind all outputs together
bind_rows(splice_output_list) -> full_splice_out
full_splice_out %>% 
  rename("CHROM" = "#CHROM") -> full_splice_out

#order chromosomes
ordered_full <- full_splice_out[order(full_splice_out[,1]),]

```


## Parse files
Take INFO column and assign each probability & position to a new column
```{r}
#for those rows with more than one gene in field INFO: keeps extra genes in "pos donor loss" column
ordered_full %>% 
  mutate(split = map(INFO, ~ str_split(.x, "[|,]")[[1]])) %>% 
  separate(split, c("INFO", "symbol", "score acc gain", "score acc loss", "score donor gain", "score donor loss", "pos acc gain", "pos acc loss", "pos donor gain", "pos donor loss"), ",") -> split_df

#write.csv(split_df, "split_df.csv", quote = FALSE)


#dont use gsub, can only take a vector at a time. vectorized approach: 
#remove unwanted symbols
split_df_ns = 
split_df %>% 
  mutate_all(funs(str_remove_all(.,"[c()\"]")))

test2 = 
  test2 %>% 
    mutate_all(funs(str_remove_all(.,"[c()\"]")))


head(test2)
```


Create dataframe for each category (i.e. 1 splice, 2 splice, 3 splices) and combine them 
```{r}
## add tag 1, 2 or 3 to know how many rows each df should have
# "," separates the output probabilities for each gene
# problem: only considers 3 alternative snps
ordered_full %>% 
   mutate(
    nr_snps = case_when(
      str_count(INFO, ",") == 0 ~  1,
      str_count(INFO, ",") == 1 ~ 2, 
      str_count(INFO, ",") == 2 ~ 3,
  )) -> full_tag


#######dataframe for 1 snp
full_tag %>% 
  filter(nr_snps == 1) %>%
  #split info
  mutate(split = map(INFO, ~ str_split(.x, "[|,]")[[1]])) %>% 
  #separate into columns
  separate(split, c("INFO", "symbol", "score_acceptor_gain", "score_acceptor_loss", "score_donor_gain", "score_donor_loss", "position_acceptor_gain", "position_acceptor_loss", "position_donor_gain", "position_donor_loss"), ",") %>%  
  #remove symbols
  mutate_all(funs(str_remove_all(.,"[c()\"]"))) %>% 
  select(-nr_snps) -> split_df1

######dataframe for 2 snp
full_tag %>% 
  filter(nr_snps == 2) %>% 
  #split info 
  mutate(split = map(INFO, ~ str_split(.x, "[|,]")[[1]])) %>% 
  #separate into columns
  separate(split, c("INFO", "symbol", "score_acceptor_gain", "score_acceptor_loss", "score_donor_gain", "score_donor_loss", "position_acceptor_gain", "position_acceptor_loss", "position_donor_gain", "position_donor_loss", "INFO_2", "symbol_2", "score_acceptor_gain_2", "score_acceptor_loss_2", "score_donor_gain_2", "score_donor_loss_2", "position_acceptor_gain_2", "position_acceptor_loss_2", "position_donor_gain_2", "position_donor_loss_2"), ",") %>%  
  #remove symbols
  mutate_all(funs(str_remove_all(.,"[c()\"]"))) %>% 
  select(-nr_snps)-> split_df2

# get the probabilities in columns labeled as "_2" and store in separate df
# remove "_2" label so that we can bind with rest of df 
split_df2 %>% 
  select(CHROM, POS, ID, REF, ALT , QUAL, FILTER, contains("2")) %>% 
  rename(setNames(names(.), gsub("_2", "", names(.)))) -> split_df2_extra 

#drop columns "_2" from df
split_df2 %>% 
  select(-contains("2"))-> split_df2

#bind
split_df2 = rbind(split_df2, split_df2_extra)


######dataframe for 3 snp
full_tag %>% 
  filter(nr_snps == 3) %>% 
   #split info 
   mutate(split = map(INFO, ~ str_split(.x, "[|,]")[[1]])) %>% 
   #separate into columns
   separate(split, c("INFO", "symbol", "score_acceptor_gain", "score_acceptor_loss", "score_donor_gain", "score_donor_loss", "position_acceptor_gain", "position_acceptor_loss", "position_donor_gain", "position_donor_loss", "INFO_2", "symbol_2", "score_acceptor_gain_2", "score_acceptor_loss_2", "score_donor_gain_2", "score_donor_loss_2", "position_acceptor_gain_2", "position_acceptor_loss_2", "position_donor_gain_2", "position_donor_loss_2", "INFO_3", "symbol_3", "score_acceptor_gain_3", "score_acceptor_loss_3", "score_donor_gain_3", "score_donor_loss_3", "position_acceptor_gain_3", "position_acceptor_loss_3", "position_donor_gain_3", "position_donor_loss_3"), ",") %>%  
  #remove symbols
  mutate_all(funs(str_remove_all(.,"[c()\"]"))) %>% 
  select(-nr_snps) -> split_df3

split_df3 %>% 
  select(CHROM, POS, ID, REF, ALT , QUAL, FILTER, contains("3")) %>% 
  rename(setNames(names(.), gsub("_3", "", names(.))))-> split_df3_3_extra

split_df3 %>% 
  select(CHROM, POS, ID, REF, ALT , QUAL, FILTER, contains("2")) %>% 
  rename(setNames(names(.), gsub("_2", "", names(.))))-> split_df3_2_extra

split_df3 %>% 
  select(-contains("2")) %>% 
  select(-contains("3"))-> split_df3

rbind(split_df3, split_df3_2_extra) -> split_df3_1_2 
rbind(split_df3_1_2, split_df3_3_extra) -> split_df3

#bind all df3

#bind all dfs
rbind(split_df1, split_df2, fill=FALSE) -> split_1_2
rbind(split_1_2, split_df3, fill=FALSE) -> full_split
```
