---
title: "Compare output for SNPs with and without RNA-Seq data associated"
output:
  html_document:
    df_print: paged
---

## Load libraries

```{r, message=FALSE, include=T}
library(tidyverse)
library(readr)
library(data.table)
library(pROC)
library(PRROC)
library(hrbrthemes)
library(viridis)
library(MLeval)
select = dplyr::select
rename = dplyr::rename
```

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = "/Users/castilln/Desktop/thesis")
```


## Load parsed output
```{r}
##SPLICEAI PARSED OUTPUT
data = fread("localdata/spliceai/output/spliceAI_out.csv")
##CCLE MUTATIONS DATA
CCLE = fread("localdata/depmap/CCLE_info")
#CCLE RNASEQ DATA
rna_seq_reads = fread("localdata/depmap/CCLE_RNAseq_reads.csv")
```

## Create unique ID for each SNP 
```{r}
##UNIQUE VARIABLE FOR SPLICEAI OUT
udata = data[, id_var := paste(ID, SYMBOL, POS, REF, ALT, sep = "_"), by = .(ID, SYMBOL, POS, REF, ALT)]

##UNIQUE VARIABLE FOR CCLE DATA
uccle = CCLE[, id_var := paste(stripped_cell_line_name, Hugo_Symbol, Start_position, Reference_Allele, Tumor_Seq_Allele1, sep = "_"), by = .(stripped_cell_line_name, Hugo_Symbol, Start_position, Reference_Allele, Tumor_Seq_Allele1)]
```


## Threshold
Find max probability 
```{r, warning=FALSE}
###CREATE A NEW COLUMN WITH THE HIGHEST PROBABILITY

##TURN SCORE COLUMNS INTO NUMERIC VARIABLES
#SELECT SCORE COLUMNS
score_col = 
  udata %>% 
  select(contains("SCORE")) 

#SELECT COLUMN NAMES
cols.num <- colnames(score_col)

#CONVERT INTO NUMERIC
udata = as.data.frame(udata)
udata[cols.num] <- sapply(udata[cols.num],as.numeric)

##SELECT MAX PROBABILITY FROM SPLICEAI OUTPUT
max_data =
  udata %>% 
  mutate(MAX_PROBABILITY = apply(X = data[,10:13], MARGIN = 1, FUN = max)) %>% 
  drop_na(MAX_PROBABILITY)

```


## Annotation
Let's see which mutations had already been annotated as splice variants in the CCLE
```{r}
## GET LIST OF ALL VARIANT ANNOTATIONS
CCLE %>% 
  pull(Variant_Classification) %>% 
  unique()

## GET ANNOTATION COLUMNS
ann_ccle = 
uccle %>%  
  select(Hugo_Symbol, CCLE_Name, Variant_Classification, Variant_annotation, id_var, primary_disease)

## FILTER FOR THOSE CLASSIFIED AS ALTERING A SPLICE SITE
CCLE_splice = 
  CCLE %>% 
    filter(Variant_Classification == "Splice_Site")
```

Join CCLE annotation with spliceAI output
```{r}
## JOIN BY HUGO SYMBOL AND CELL LINE NAME
ann_data = 
  max_data %>% 
  left_join(ann_ccle, by = "id_var") %>% 
  select(-c("Hugo_Symbol", "QUAL", "FILTER", "CCLE_Name")) #drop unnecessary variables
```

## Prepare data for ROC curves
```{r}
## GIVE SPLICE SITE 1 AND EVERYTHING ELSE 0 
roc_data = 
  ann_data %>% 
  mutate(splice_altering = case_when(
    Variant_Classification == "Splice_Site" ~ 1, 
    Variant_Classification != "Splice_Site" ~ 0
  )) %>% 
  mutate(MAX_PROBABILITY = as.numeric(MAX_PROBABILITY)) %>% 
  drop_na()

head(roc_data)
```

## Check spliceAI output for data with and w/o RNA-seq
Get unique ID for those observation w/o RNA-seq
```{r}
rna_seq_reads = 
  rna_seq_reads %>% 
  rename("Tumor_Sample_Barcode" = "V1")

no_rnaseq = 
  CCLE %>% 
  anti_join(rna_seq_reads, by = "Tumor_Sample_Barcode")

##CREATE UNIQUE ID TO BE ABLE TO COMPARE WITH SPLICE AI OUTPUT
u_no_rnaseq = no_rnaseq[, id_var := paste(stripped_cell_line_name, Hugo_Symbol, Start_position, Reference_Allele, Tumor_Seq_Allele1, sep = "_"), by = .(CCLE_Name, Hugo_Symbol, Start_position, Reference_Allele, Tumor_Seq_Allele1)]
```

Generate new variable with YES/NO info for RNAseq data 
```{r}
data_no_rnaseq = 
  roc_data %>% 
  semi_join(u_no_rnaseq, by = "id_var") %>% 
  mutate(rnaseq = "NO") %>% 
  mutate(splice_altering = as.factor(splice_altering))

data_rnaseq = 
  roc_data %>% 
  anti_join(u_no_rnaseq, by = "id_var") %>% 
  mutate(rnaseq = "YES") %>% 
  mutate(splice_altering = as.factor(splice_altering))
```

## Plot
Function to produce data statistics
```{r}
data_summary <- function(x) {
   m <- mean(x)
   ymin <- m-sd(x)
   ymax <- m+sd(x)
   return(c(y=m,ymin=ymin,ymax=ymax))
}
```

```{r, fig.width=7, fig.height=6, warning=F, message=FALSE}
##VIOLIN PLOT 
ggplot(data = data_rnaseq, aes(x = splice_altering, y = MAX_PROBABILITY, fill = splice_altering)) +
  geom_violin(trim = F) + 
  #LOG TRANSFORM PROBABILITIES TO OBTAIN BETTER DENSITY FOR VISUALIZATION (CHANGED Y AXIS LABEL ACCORDINGLY)
  scale_y_log10() + 
  facet_wrap(facets = "primary_disease") +
  scale_fill_viridis(discrete=T, name="Annotated as splice-altered variants based on RNA-Seq data", labels = c("NO", "YES")) +
  scale_x_discrete(labels=c("0" = "NO", "1" = "YES")) +
  theme_bw() +
  theme( 
        axis.ticks.x = element_blank(),
        legend.position = "bottom",
        plot.title = element_text(face = "bold"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.text.x = element_blank()
        ) +
  ggtitle("SNPs for which RNA-seq data available") + 
  ylab("Predicted log probability of being a splice altering mutation") + 
  xlab("") +
  stat_summary(fun.data = data_summary)

```

```{r, include=FALSE}
ggsave("spliceai_rnaseq_viz_probabilities.png", width = 10, height = 10)
```


No RNA-seq
```{r, fig.width=7, fig.height=6, warning=F, message=FALSE}
##VIOLIN PLOT 
ggplot(data = data_no_rnaseq, aes(x = splice_altering, y = MAX_PROBABILITY, fill = splice_altering)) +
  geom_violin(trim = F) + 
  #LOG TRANSFORM PROBABILITIES TO OBTAIN BETTER DENSITY FOR VISUALIZATION (CHANGED Y AXIS LABEL ACCORDINGLY)
  scale_y_log10() + 
  facet_wrap(facets = "primary_disease") +
  scale_fill_viridis(discrete=T, name="Annotated as splice-altered variants based on RNA-Seq data", labels = c("NO", "YES")) +
  scale_x_discrete(labels=c("0" = "NO", "1" = "YES")) +
  theme_bw() +
  theme( 
        axis.ticks.x = element_blank(),
        legend.position = "bottom",
        plot.title = element_text(face = "bold"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.text.x = element_blank()
        ) +
  ggtitle("SNPs without RNA-seq data available") + 
  ylab("Predicted log probability of being a splice altering mutation") + 
  xlab("") +
  stat_summary(fun.data = data_summary)
```

