---
title: "SpliceAI analysis - Variant classification"
output:
  html_document:
    df_print: paged
---

## Load libraries

```{r, message=FALSE, include=FALSE}
library(tidyverse)
library(readr)
library(data.table)
library(pROC)
library(PRROC)
library(hrbrthemes)
library(viridis)
library(ggbeeswarm)
library(MLeval)
select = dplyr::select
rename = dplyr::rename
```

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = "/Users/castilln/Desktop/thesis/")
```


## Load parsed output
```{r}
##SPLICEAI PARSED OUTPUT
data = fread("localdata/spliceai/output/spliceAI_out.csv")
##ccle_meta MUTATIONS DATA
ccle_meta = fread("localdata/depmap/ccle_info_21q1.csv")
```

Join ccle_meta with metadata
```{r}
ccle_meta = 
  ccle_meta %>% 
  select(-V1)
```

## Create unique ID for each SNP 
```{r}
##UNIQUE VARIABLE FOR SPLICEAI OUT
udata = data[, id_var := paste(ID, SYMBOL, POS, REF, ALT, sep = "_"), by = .(ID, SYMBOL, POS, REF, ALT)]

##UNIQUE VARIABLE FOR ccle_meta DATA
uccle_meta = ccle_meta[, id_var := paste(stripped_cell_line_name, Hugo_Symbol, Start_position, Reference_Allele, Tumor_Seq_Allele1, sep = "_"), by = .(stripped_cell_line_name, Hugo_Symbol, Start_position, Reference_Allele, Tumor_Seq_Allele1)]

head(uccle_meta)
```


## Threshold
Find max probability 
```{r, warning=FALSE}
###CREATE A NEW COLUMN WITH THE HIGHEST PROBABILITY

##TURN SCORE COLUMNS INTO NUMERIC VARIABLES
#SELECT SCORE COLUMNS
score_col = 
  udata %>% 
  select(contains("SCORE")) 

#SELECT COLUMN NAMES
cols.num <- colnames(score_col)

#CONVERT INTO NUMERIC
udata = as.data.frame(udata)
udata[cols.num] <- sapply(udata[cols.num],as.numeric)

##SELECT MAX PROBABILITY FROM SPLICEAI OUTPUT
max_data =
  udata %>% 
  mutate(MAX_PROBABILITY = apply(X = data[,10:13], MARGIN = 1, FUN = max)) %>% 
  drop_na(MAX_PROBABILITY)

```


## Annotation
Let's see which mutations had already been annotated as splice variants in the ccle_meta
```{r}
## GET LIST OF ALL VARIANT ANNOTATIONS
ccle_meta %>% 
  pull(Variant_Classification) %>% 
  unique()

## GET ANNOTATION COLUMNS
ann_ccle_meta = 
uccle_meta %>%  
  select(Hugo_Symbol, stripped_cell_line_name, Variant_Classification, Variant_annotation, id_var, primary_disease)


## FILTER FOR THOSE CLASSIFIED AS ALTERING A SPLICE SITE
ccle_meta_splice = 
  ccle_meta %>% 
    filter(Variant_Classification == "Splice_Site")
```

Join ccle_meta annotation with spliceAI output
```{r}
## JOIN BY HUGO SYMBOL AND CELL LINE NAME
ann_data = 
  max_data %>% 
  left_join(ann_ccle_meta, by = "id_var") %>% 
  select(-c("Hugo_Symbol", "QUAL", "FILTER", "stripped_cell_line_name")) #drop unnecessary variables
```


## Prepare data for ROC curves
```{r}
## GIVE SPLICE SITE 1 AND EVERYTHING ELSE 0 
roc_data = 
  ann_data %>% 
  mutate(splice_altering = case_when(
    Variant_Classification == "Splice_Site" ~ 1, 
    Variant_Classification != "Splice_Site" ~ 0
  )) %>% 
  mutate(MAX_PROBABILITY = as.numeric(MAX_PROBABILITY)) %>% 
  drop_na() %>% 
  mutate(PREDICTED = case_when(
    MAX_PROBABILITY >= 0.5 ~ 1, 
    MAX_PROBABILITY < 0.5 ~ 0 
  ))

head(roc_data)
```

## ROC curve
AUC: Area Under the Curve
```{r}
rocobj <- roc(response = roc_data$splice_altering, predictor = roc_data$MAX_PROBABILITY, partial.auc.correct = TRUE, percent = TRUE)

## PLOT 
plot.roc(rocobj, print.auc = TRUE)

# Extract AUC
AUC_roc = as.double(auc(rocobj))/100

p_roc = 
  ggroc(rocobj, colour = "black") +
  geom_segment(aes(x =100, xend = 0, y = 0, yend = 100),color = "grey") +
  annotate("text", x = 25, y = 25, label = paste0("AUC = ", round(AUC_roc,2))) +
  theme_bw() +
  xlim(100,0) +
  ylim(0,100) +
  xlab("Specificity") +
  ylab("Sensitivity") +
  theme(aspect.ratio = 1)

p_roc
```

## PRECISION RECALL CURVE

```{r}
pr <- pr.curve(roc_data$MAX_PROBABILITY,roc_data$splice_altering, curve = T) 

# Extract and format data
pr_data = pr$curve %>% 
  as_tibble()
colnames(pr_data) = c("Recall","Precision","Color")
set.seed(1337)
pr_data = sample_n(pr_data,1000) #subsample for easier plotting
pr_data = 
  pr_data %>% 
  mutate(Recall = 100*Recall,
         Precision = 100*Precision)
pr_data = rbind(pr_data,c(0,0,0))

# Add AUC
AUC_pr = pr$auc.integral

p_pr = 
  pr_data %>% 
  ggplot(aes(y=Precision,x=Recall)) +
  geom_line() +
  geom_segment(aes(x =0, xend = 100, y = 0, yend = 0),color = "grey") +
  theme_bw() +
  annotate("text", x = 75, y = 25, label = paste0("AUC = ", sprintf("%.2f",round(AUC_pr,3)))) +
  #xlim(100,0) +
  #ylim(0,100) +
  theme(aspect.ratio = 1)
  
p_pr
```

Combine plots
```{r}
library(patchwork)
p_combined_auc = p_roc + p_pr
p_combined_auc
```
Save plots
```{r}
ggsave(plot = p_combined_auc,file = "figures/results/spliceai/roc_pr.pdf")
```



## Contingency table
```{r}
cont_data = 
  roc_data %>% 
  mutate(predictor = case_when(
    MAX_PROBABILITY >= 0.5 ~ "Max prob >= 0.5",
    MAX_PROBABILITY < 0.5 ~ "Max prob < 0.5"
  )) %>% 
  mutate(pred_splice_altering = case_when(
    splice_altering == 1 ~ "Not annotated", 
    splice_altering == 0 ~ "Annotated splice altering"
  ))
```


```{r}
cont = table(distinct(cont_data)$predictor,distinct(cont_data)$pred_splice_altering)

cont[2:1, 1:2]
```


## Violin plots
Function to produce data statistics
```{r}
data_summary <- function(x) {
   m <- mean(x)
   ymin <- m-sd(x)
   ymax <- m+sd(x)
   return(c(y=m,ymin=ymin,ymax=ymax))
}
```


Data distribution
```{r, fig.width=3, fig.height=2}
##BOXPLOTS  
roc_data = 
  roc_data %>% 
    mutate(splice_altering = as.factor(splice_altering)) 


ggplot(data = roc_data, aes(x = splice_altering, y = MAX_PROBABILITY, fill = splice_altering)) +
  geom_violin(trim = F) + 
  scale_y_log10() +
  scale_fill_manual(values = c("0" = "gray", "1" = "yellow")) + 
  theme_bw() +
  theme( 
        axis.ticks.x = element_blank(),
        legend.position = "none",
        plot.title = element_text(face = "bold"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        ) +
  scale_x_discrete(labels=c("0" = "NO", "1" = "YES")) +
  geom_hline(yintercept=0.5, linetype="dashed", color = "red") +
  ylab("Probability of splice altering mutation") + 
  xlab("Splice altering") +
  stat_summary(fun.data = data_summary)

ggsave("figures/results/spliceai/splice_altering_general.png", height = 3, width = 3)
```


## Analyze prediction for each disease

```{r, fig.width=7, fig.height=6, warning=FALSE}
## EXCLUDE DATA FROM UNKNOWN OR NOT-CANCEROUS CELL LINES AND FROM ADRENAL CANCER (1 CELL LINE)
data_all_cancers = 
  select(filter(roc_data, primary_disease != "Unknown" & primary_disease != "Non-Cancerous" & primary_disease != "Adrenal Cancer"), c("primary_disease", "MAX_PROBABILITY", "splice_altering")) %>% 
  mutate(splice_altering = as.factor(splice_altering))

## MAKE SURE SPLICE_ALTERING IS A FACTOR
ggplot(data = data_all_cancers, aes(x = splice_altering, y = MAX_PROBABILITY, fill = splice_altering)) +
  geom_violin(trim = F) +
  #geom_quasirandom(method = "pseudorandom",alpha = 0.5, size = 0.001) +
  scale_y_log10() +
  facet_wrap(facets = "primary_disease") +
  #facet_grid(. ~ primary_disease) +
  scale_fill_manual(values = c("0" = "gray", "1" = "yellow"), name="Annotated as splice-altered variants based on RNA-Seq data", labels = c("NO", "YES")) +
  theme_bw() +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(),
        legend.position = "bottom",
        plot.title = element_text(face = "bold"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        ) +
  scale_x_discrete(labels=c("0" = "NO", "1" = "YES")) +
  ylab("Predicted log probability of being a splice altering mutation") + 
  xlab("") +
  geom_hline(yintercept=0.5, linetype="dashed", color = "red") +
  stat_summary(fun.data = data_summary)

ggsave("figures/results/spliceai/splice_altering_cancer.png", width = 11, height = 8)
```

```{r, include=FALSE}
#ggsave("figures/results/spliceai/spliceai_viz_probabilities_jitter.png", width = 10, height = 10)
```


## Set  threshold and prepare table to present data
Select information of interest from ccle_meta_splice
```{r}
meta = 
  uccle_meta %>% 
  select(c("DepMap_ID", "Annotation_Transcript", "id_var"))
```

Threshold at 0.5 

```{r}
ths5 = 
  ann_data %>% 
  filter(MAX_PROBABILITY >= 0.5) %>% 
  select(-c("INFO", "Variant_annotation")) %>%
  left_join(meta, by = "id_var") %>% 
  relocate(DepMap_ID)

#GET LIST FROM ccle_meta
list_ccle_meta = 
  ccle_meta %>% 
  select(DepMap_ID, stripped_cell_line_name, primary_disease) %>% 
  distinct() %>% 
  rename("CELL_LINE_NAME" = "stripped_cell_line_name")

##ANNOTATE OFF TARGET SPLICE (i.e. if NA in DepMap_ID
ths5_off = 
  ths5 %>% 
  mutate(off_target = ifelse(is.na(DepMap_ID), 1, 0)) %>% 
  rename("CELL_LINE_NAME" = "ID") %>% 
  select(-c("DepMap_ID", "primary_disease")) %>% 
  left_join(list_ccle_meta, by = "CELL_LINE_NAME") %>% 
  relocate(DepMap_ID, .before = CHROM)

ths5_off %>% 
  filter(off_target == 1) %>% 
  dim()
```

```{r}
#INVESTIGATE REMAINING NA 
na_depmap = 
  is.na(ths5_off$DepMap_ID)

why_na = ths5_off[na_depmap,]
```

```{r}
saveRDS(ths5_off, file="github/results/spliceAI/ths5.rds")
```

Threshold at 0.8 (max probability)
```{r}
ths8 = 
  ann_data %>% 
  filter(MAX_PROBABILITY >= 0.8) %>% 
  select(-c("INFO", "Variant_annotation")) %>%
  left_join(meta, by = "id_var") %>% 
  relocate(DepMap_ID)


head(ths8)
```

```{r}
saveRDS(ths8, file="github/results/spliceAI/ths8.rds")
```


Filter those SNPs without DepMap ID
```{r}
noID = 
  ths8 %>% 
  filter(is.na(DepMap_ID)) %>% 
  select(-c("Variant_Classification", "primary_disease", "Annotation_Transcript")) %>% 
  rename("CELL_LINE" = "ID")
```

Generate a list of primary disease based on cell line name
```{r}
disease = 
  ccle_meta %>% 
  select(c("stripped_cell_line_name", "primary_disease")) %>% 
  distinct() %>% 
  rename("CELL_LINE" = "stripped_cell_line_name", 
         "DISEASE" = "primary_disease")
```

Add disease information to SNPs w/o DepMapID
```{r}
noID = 
  noID %>% 
  left_join(disease, by = "CELL_LINE") %>% 
  arrange(desc(MAX_PROBABILITY))
```

```{r}
head(noID)
```


Get SNPs in leukemia cell lines
```{r}
noID %>% 
  filter(DISEASE.x  == "Pancreatic Cancer")
```
Let's try to relate each new predicted SNP with the SNP from ccle_meta
```{r}
ccle_meta %>% 
  filter(stripped_cell_line_name == "L33")
```


Let's try to verify that those alterations predicted by spliceAI that lack a DepMap ID are not present in the mutations df
```{r}
head(ccle_meta_mutations)
```



