---
title: "SpliceAI results analysis excluding exons"
output:
  html_document:
    df_print: paged
---

## Load libraries

```{r, message=FALSE, include=T, echo=T}
library(tidyverse)
library(readr)
library(data.table)
library(pROC)
library(PRROC)
library(MLeval)
select = dplyr::select
rename = dplyr::rename
```

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = "/Users/castilln/Desktop/thesis")
```


## Load parsed output

```{r, message=FALSE, include=T, echo=T, warning=F}
##SPLICEAI PARSED OUTPUT
data = fread("localdata/spliceai/output/spliceAI_out.csv")
##CCLE MUTATIONS DATA
CCLE = fread("localdata/depmap/CCLE_info")
```

## Create unique ID for each SNP 
```{r}
##UNIQUE VARIABLE FOR SPLICEAI OUT
udata = data[, id_var := paste(ID, SYMBOL, POS, REF, ALT, sep = "_"), by = .(ID, SYMBOL, POS, REF, ALT)]

##UNIQUE VARIABLE FOR CCLE DATA
uccle = CCLE[, id_var := paste(stripped_cell_line_name, Hugo_Symbol, Start_position, Reference_Allele, Tumor_Seq_Allele1, sep = "_"), by = .(stripped_cell_line_name, Hugo_Symbol, Start_position, Reference_Allele, Tumor_Seq_Allele1)]
```


## Threshold
Find max probability predicted by SpliceAI
```{r, warning=FALSE}
###CREATE A NEW COLUMN WITH THE HIGHEST PROBABILITY

##TURN SCORE COLUMNS INTO NUMERIC VARIABLES
#SELECT SCORE COLUMNS
score_col = 
  udata %>% 
  select(contains("SCORE")) 

#SELECT COLUMN NAMES
cols.num <- colnames(score_col)

#CONVERT INTO NUMERIC
udata = as.data.frame(udata)
udata[cols.num] <- sapply(udata[cols.num],as.numeric)

##SELECT MAX PROBABILITY FROM SPLICEAI OUTPUT
max_data =
  udata %>% 
  mutate(MAX_PROBABILITY = apply(X = data[,10:13], MARGIN = 1, FUN = max)) %>% 
  drop_na(MAX_PROBABILITY)

```

## Annotation
Let's see which mutations had already been annotated as splice variants in the CCLE
```{r}
## GET LIST OF ALL VARIANT ANNOTATIONS
CCLE %>% 
  pull(Variant_Classification) %>% 
  unique()

## GET ANNOTATION COLUMNS
ann_ccle = 
uccle %>%  
  select(Hugo_Symbol, CCLE_Name, Variant_Classification, Variant_annotation, id_var, primary_disease)

## FILTER FOR THOSE CLASSIFIED AS ALTERING A SPLICE SITE
CCLE_splice = 
  CCLE %>% 
    filter(Variant_Classification == "Splice_Site")
```

Join CCLE annotation with spliceAI output
```{r}
## JOIN BY HUGO SYMBOL AND CELL LINE NAME
ann_data = 
  max_data %>% 
  left_join(ann_ccle, by = "id_var") %>% 
  select(-c("Hugo_Symbol", "QUAL", "FILTER", "CCLE_Name")) #drop unnecessary variables
```

## Analysis excluding exons
Find mutations that do not occur in exons
```{r}
#cDNA change column: changes in coding DNA
non_coding = 
  CCLE %>% 
  filter(is.na(cDNA_Change))

## ADD UNIQUE ID  
id_non_coding = non_coding[, id_var := paste(CCLE_Name, Hugo_Symbol, Start_position, Reference_Allele, Tumor_Seq_Allele1, sep = "_"), by = .(CCLE_Name, Hugo_Symbol, Start_position, Reference_Allele, Tumor_Seq_Allele1)]

## GET ANNOTATION COLUMN
ann_non_coding = 
id_non_coding %>%  
  select(Hugo_Symbol, CCLE_Name, Variant_Classification, Variant_annotation, id_var, primary_disease)
```

Join CCLE annotation with spliceAI output
```{r}
## JOIN BY HUGO SYMBOL AND CELL LINE NAME
ann_non_coding = 
  max_data %>% 
  left_join(ann_ccle, by = "id_var") %>% 
  select(-c("Hugo_Symbol", "QUAL", "FILTER", "CCLE_Name")) #drop unnecessary variables
```

## Prepare data for ROC curves

```{r}
## GIVE SPLICE SITE 1 AND EVERYTHING ELSE 0 
roc_non_coding = 
  ann_non_coding %>% 
  mutate(splice_altering = case_when(
    Variant_Classification == "Splice_Site" ~ 1, 
    Variant_Classification != "Splice_Site" ~ 0
  )) %>% 
  mutate(MAX_PROBABILITY = as.numeric(MAX_PROBABILITY)) %>% 
  drop_na()

head(roc_non_coding)
```

## ROC curve

AUC: Area Under the Curve
```{r}
## PREDICTOR = spliceAI out
## RESPONSE = Variant_Classification

rocobj_nc <- roc(response = roc_non_coding$splice_altering, predictor = roc_non_coding$MAX_PROBABILITY, partial.auc.correct = TRUE, percent = TRUE)

## PLOT 
plot.roc(rocobj_nc, print.auc = TRUE)

```

## PRECISION RECALL CURVE

```{r}
pr_nc <- pr.curve(roc_non_coding$MAX_PROBABILITY,roc_non_coding$splice_altering, curve = T, max.compute = TRUE, 
  min.compute = TRUE, rand.compute = TRUE)
plot(pr_nc, max.plot = TRUE, min.plot = TRUE, rand.plot = TRUE, fill.area = TRUE)
```
