---
title: "Random forest - classification spliceosome mutated cell lines based on dependency of genes"
output: html_notebook
---

```{r}
library(tidyverse)
library(data.table)
library(randomForest)
```

Set wd 
```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = "/Users/castilln/Desktop/thesis/localdata")
```

Load data - transcription profile & dependencies
```{r}
transcripts = read_csv("depmap/CCLE_RNAseq_transcripts.csv")

#METADATA
sample_info <- read_csv("depmap/sample_info.csv") 

#LIST OF MUTATIONS FROM CCLE MUTATIONS THAT AFFECT THE SPLICEOSOME
mutations_spliceosome = fread("depmap/mutations_spliceosome.csv")

#DEPENDENCY DATA
dep_df = fread("depmap/achilles/Achilles_gene_dependency.csv")

#GENE EFFECT
effect = fread("depmap/achilles/Achilles_gene_effect.csv")

#CCLE SNP data
ccle = fread("depmap/CCLE_info")
```

Annotate cancer type
```{r}
cancer_list = 
  ccle %>% 
  select(c("DepMap_ID", "primary_disease", "stripped_cell_line_name")) %>% 
  distinct()


ann_df = 
  effect %>% 
  left_join(cancer_list, by = "DepMap_ID")
```

Annotate spliceosome
```{r}
ann_spliceosome = 
  ann_df %>% 
  mutate(spliceosome_mutated = case_when(
     ann_df$stripped_cell_line_name %in% mutations_spliceosome$stripped_cell_line_name ~ "YES", #if the mutation is in a gene from the spliceosome: 1 
            !ann_df$stripped_cell_line_name %in% mutations_spliceosome$stripped_cell_line_name ~ "NO")) 


head(ann_df)
```

```{r}
pretrain_df = 
  ann_spliceosome %>% 
  drop_na(primary_disease) %>% 
  select(-stripped_cell_line_name)

head(pretrain_df)
```

NA imputation - let√§s guess that those genes with NA are not cancer dependencies
```{r}
pretrain_df[is.na(pretrain_df)] <- 0
```

Random forest - do not run using formula; make sure response variable y is transformed to factors
```{r}
train_x = 
  pretrain_df %>%  
  select(-c(spliceosome_mutated, DepMap_ID))

train_y = 
  factor(pretrain_df$spliceosome_mutated)

#CHECK NUMBER OF TREES
dep.rf = 
  randomForest(train_x, train_y, importance = TRUE, proximity = TRUE, ntree = 1000)
```

Plot importance of predictor variables
```{r, fig.height=7}
varImpPlot(dep.rf)
```

```{r}
print(dep.rf)
print(importance(dep.rf, type=2))
```

Train random forest with 10000 trees 
```{r}
dep.rf_10000 = 
  randomForest(train_x, train_y, importance = TRUE, proximity = TRUE, ntree = 10000)
```

```{r, fig.height=7}
varImpPlot(dep.rf_10000)
```

```{r}
print(dep.rf_10000)
print(importance(dep.rf_10000, type=2))
```
